addon/content/es-modules/contact-manager.js

```diff
@@ -0,0 +1,175 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import { freshColor } from "./utils.js";
+
+class ContactManager {
+  constructor() {
+    this._cache = new Map();
+    this._colorCache = new Map();
```

Comment:
Just remembered, we should also listen out for changes in the address book and update these as necessary. There's some [example code here](https://github.com/thunderbird-conversations/thunderbird-conversations/compare/8183f45...Standard8:contacts#diff-cc0d0493d0507f6c06da7e674bb330d18d985f6cd6e797ab748152583ddf7fe2R15-R59).

We probably get away with it more on the current code since the contacts are cached for just the life of stub.xhtml (which changes on folder change), rather than the entire session. Though I think we should be handling it anyway.

I'm not too worried about updating the display straight away at the moment though.

We possibly should also consider a later follow-up to limit the cache sizes - I'm not sure who, but I guess some people might be seeing emails from lots of people.